### PR TITLE
Fix #2770: Implement Intl.getCanonicalLocales

### LINKED ISSUE #2770: Implement Intl.getCanonicalLocales
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/getCanonicalLocales

https://tc39.github.io/ecma402/#sec-intl.getcanonicallocales
https://tc39.github.io/ecma402/#sec-canonicalizelocalelist

### PR DESCRIPTION
**Pull request by dilijev**
https://tc39.github.io/ecma402/#sec-intl.getcanonicallocales
https://tc39.github.io/ecma402/#sec-canonicalizelocalelist

See:
https://github.com/tc39/ecma402/pull/78
https://github.com/tc39/ecma402/issues/46

Fixes #2770.

### CODE REVIEW

**Comment by dilijev** [comment]
@dotnet-bot test Ubuntu static_ubuntu_linux_release please

**Comment by dilijev** [comment]
Ping @bterlson @kunalspathak @tcare

File: test/Intl/GetCanonicalLocales.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            //      5. Let len be ? ToLength(? Get(O, "length")).
+            //      6. Let k be 0.
+            //      7. Repeat, while k < len
+            // Since ToLength(undefined) === 0, we don't enter the loop (essentially treat the input as a zero-length array).
+            //      ToLength(undefined) -> ToInteger(undefined) -> ToNumber(undefined) -> NaN
+            //      ToInteger converts NaN to +0.
+            assert.areEqual(Intl.getCanonicalLocales({ '0': 'en-us' }), [], "Objects which might look like arrays are fine, but treated as 0 length.");
+            assert.areEqual(Intl.getCanonicalLocales({ 'a': 'b' }), [], "Arbitrary Objects are fine, treated as 0-length arrays.");
+
+            // TODO test this
```
Comment (tcare) [inline test/Intl/GetCanonicalLocales.js]: Are we leaving this?

File: test/Intl/GetCanonicalLocales.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        // Therefore, anything that fits into the general language tag grammar should be canonicalized.
+        // * ECMA 402 #sec-isstructurallyvalidlanguagetag
+        // * (Note: The above basically just refers to RFC 5646 section 2.1)
+        body: function () {
+            assert.areEqual(Intl.getCanonicalLocales('en-zz'), ['en-ZZ'], "en-ZZ: English as used in [unsupported locale ZZ]");
+            assert.areEqual(Intl.getCanonicalLocales('ZZ-us'), ['zz-US'], "zz-US: [unsupported language zz] as used in US");
+            assert.areEqual(Intl.getCanonicalLocales('xx-zzz'), ['xx-zzz'], ""); // TODO double-check canonicalization routine, v8 makes this 'zzz'
+            assert.areEqual(Intl.getCanonicalLocales('xx-abcd-zz'), ['xx-Abcd-ZZ'],
+                "xx-Abcd-ZZ: [unsupported language xx] using [unsupported script Abcd] as used in [unsupported locale ZZ]");
+            // TODO -u- extension keys
```
Comment (tcare) [inline test/Intl/GetCanonicalLocales.js]: +1

File: test/Intl/GetCanonicalLocales.js (hunk 1 above)
Comment (dilijev) [inline test/Intl/GetCanonicalLocales.js]: Yeah sorry about the TODOs. I wanted to get this out for review. I'll add more tests to cover the gaps in the next iteration.

File: lib/Runtime/Library/EngineInterfaceObject.cpp [hunk 3]
```diff
@@ -1,4 +1,4 @@
- //-------------------------------------------------------------------------------------------------------
+//-------------------------------------------------------------------------------------------------------
```
Comment (bterlson) [inline lib/Runtime/Library/EngineInterfaceObject.cpp]: There should be awards for bug fixes this severe.

Comment (dilijev) [inline lib/Runtime/Library/EngineInterfaceObject.cpp]: lol. Like I said, drive by cleanup. :P

File: test/Intl/DateTimeFormatOptions.js [hunk 4]
```diff
@@ -51,7 +51,7 @@ var tests = [
                 }
                 catch (e) {
                     if (!(e instanceof RangeError)) {
-                        fail("Incorrect exception was thrown.");
+                        assert.fail("Incorrect exception was thrown.");
```
Comment (dilijev) [inline test/Intl/DateTimeFormatOptions.js]: /fyi @veler

**Comment by dilijev** [comment]
Note: I'm porting this change to release/2.0 since it seems relevant there. This also involves porting #2726 from master to release/2.0 so I included that in this PR as well.

**Comment by dilijev** [comment]
Fixing CI in https://github.com/Microsoft/ChakraCore/pull/2962

**Comment by dilijev** [comment]
Nevermind -- after talking offline with @digitalinfinity this is not needed in release/2.0. Retargeting back to master.

**Pull request merged**